Postgres hometask

MihailNenahov opened the pull request:
finished rewriting program
need to try it with postgres, but u already can to write comments.

Commit:
initial commit

finished rewriting program
need to try it with docker and postgres

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+package memory
```
Package called memory because it's in memory implementation of the storage, there no need to rewrite code in this file, you should write code in separate package pkg/repository/postgres to allow through dependency injection use one of implementations. For example in memory on testing and postures on production

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+func NewContactsRepositoryInMemory() *ContactsRepositoryInMemory {
+	connStr := "user=mishamba password=123 dbname=some-postgres sslmode=disable" //TODO
```
you should pass connection to database to constructor function instead of initialise it in constructor

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+
+type ContactsRepositoryInMemory struct {
+	storage *sql.DB
+	lastID  uint
```
Sql databases can generate for you Primary key automatically https://www.w3schools.com/sql/sql_autoincrement.asp

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+		if c.Phone == contact.Phone {
+			return model.Contact{}, fmt.Errorf("contact with phone %q already exists", c.Phone)
+		}
+	}
```
You don't have to read all data from database, all SQL databases support UNIQUE constraints see https://www.w3schools.com/sql/sql_unique.asp

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+
+	r.lastID++
+	contact.ID = r.lastID
+	_, err = r.storage.Exec("insert into Contacts (ID, FirstName, LastName, Phone, Email) values ($1, $2, $3, $4, $5)", contact.ID, contact.FirstName, contact.LastName, contact.Phone, contact.Email)
```
Readability for sql queries can be improved, as a convention people write all sql commands using caps lock and using snake case for naming

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+
+func (r *ContactsRepositoryInMemory) ListAll() ([]model.Contact, error) {
+	var result []model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts")
```
Naming ambiguous, row is a plural object due to this I recommend to name it ```rows```

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+	defer row.Close()
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Phone, &contact.Email)
+	if err != nil {
+		return contact, nil
```
I don't understand what the reason to return dummy data and nil error in case of not nil error from Scan

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+	defer row.Close()
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Phone, &contact.Email)
+	if err != nil {
+		return contact, nil
```
Bad error handling

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+		var tmp model.Contact
+		err := row.Scan(&tmp.ID, &tmp.FirstName, &tmp.LastName, &tmp.Phone, &tmp.Email)
+		if err != nil {
+			return result, err
```
It's better to return nothing in case of error to don't allow consumer use data which can be not valid

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+		}
+	}
+
+	r.lastID++
```
In memory realisation was concurrency save, there is a race condition in case of concurrent calls to Repository

MihailNenahov replied:
so if there will be no  `lastID`  so we wount have any concurrency. is it correct?

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+	var contact model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts where ID = ?", id)
+	if err != nil {
+		return model.Contact{}, nil
```
bad error handling

burov replied:
You should return and error with problem description

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+	var contact model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts where ID = ?", phone)
+	if err != nil {
+		return model.Contact{}, nil
```
Bad error handling

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+	var contact model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contact where Email = ?", email)
+	if err != nil {
+		return contact, err
```
It's better to return nothing In case of error, it's does not allow package consumer use data which might be invalid in case of error

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+	defer row.Close()
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Phone, &contact.Email)
+	if err != nil {
+		return contact, nil
```
Bad error handling

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+
+func (r *ContactsRepositoryInMemory) SearchByName(name string) ([]model.Contact, error) {
+	var contacts []model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts where Email = ?", name)
```
Search in memory repository allow you to search by prefix, there is strict equal condition see https://www.w3schools.com/sql/sql_like.asp

burov commented on sql/pkg/repository/memory/memory.go:
```diff
@@ -0,0 +1,162 @@
+	count, _ := info.RowsAffected()
+
+	if count > 1 {
+		return errors.New("deleted " + string(count) + " strings, but expected one")
```
It's not needed to return an error in case of 0 affected rows, the most common option to implement delete, return nil error if entity have already don't exist

burov reviewed (changes requested):
Migration is missing as well some problems in code see comments

wshaman commented on sql/cmd/cli/main.go:
```diff
@@ -0,0 +1,221 @@
+	return nil
+}
+
+func GetByID(rep model.ContactsRepository) error {
```
If this func is called like `GetByID`, it should be kind of `GetByID(id uint)`

MihailNenahov replied:
so i need to rewrite all program without using memory package?

Commit:
after reading comments

corrected err check
made queries more readable
deleted lastID field

Commit:
after reading comments

rewrote memory. now it's uses only `*sql.DB`
addedd migration files
rewrote main to use memory package

Commit:
after questions after class

added old memory package and renamed new file with database

Commit:
after reading comments

added UNIQUE Constrain
deleted redundant code(because of adding Constrain)
